src/SAWScript/MGU.hs

```diff
@@ -537,39 +541,76 @@ resolveUnificationVar pos i t =
 -- (to add to the cumulative substitution we build up) that makes them
 -- the same.
 mgu :: Type -> Type -> Either FailMGU Subst
-mgu (TyUnifyVar _ i) (TyUnifyVar _ j) | i == j = return emptySubst
-mgu (TyUnifyVar pos i) t2 = resolveUnificationVar pos i t2
-mgu t1 (TyUnifyVar pos i) = resolveUnificationVar pos i t1
-mgu r1@(TyRecord _ ts1) r2@(TyRecord _ ts2)
-  | M.keys ts1 /= M.keys ts2 =
-      failMGU "Record field names mismatch." r1 r2
-  | otherwise = case mgus (M.elems ts1) (M.elems ts2) of
-      Right result -> Right result
-      Left msgs -> Left $ failMGUAdd msgs r1 r2
-mgu c1@(TyCon _ tc1 ts1) c2@(TyCon _ tc2 ts2)
-  | tc1 == tc2 = case mgus ts1 ts2 of
-      Right result -> Right result
-      Left msgs ->
-        case tc1 of
-          FunCon -> Left $ failMGUAddFun msgs c1 c2
-          _ -> Left $ failMGUAdd msgs c1 c2
-  | otherwise = case tc1 of
-      FunCon ->
-        failMGU "Term is not a function. (Maybe a function is applied to too many arguments?)" c1 c2
-      _ ->
-        failMGU ("Mismatch of type constructors. Expected: " ++ pShow tc1 ++ " but got " ++ pShow tc2) c1 c2
-mgu (TyVar _ a) (TyVar _ b)
-  | a == b = return emptySubst
-mgu t1 t2 = failMGU "Mismatch of types." t1 t2
-
+mgu t1 t2 = case (t1, t2) of
+
+  (TyUnifyVar _ i, TyUnifyVar _ j) | i == j ->
+      -- same unification var, nothing to do
+      return emptySubst
+
+  (TyUnifyVar pos i, _) ->
+      -- one side is a unification var, resolve it
+      resolveUnificationVar pos i t2
+
+  (_, TyUnifyVar pos i) ->
+      -- one side is a unification var, resolve it
+      resolveUnificationVar pos i t1
+
+  (TyRecord _ ts1, TyRecord _ ts2) | M.keys ts1 /= M.keys ts2 ->
+      -- records with different keys
+      failMGU "Record field names mismatch." t1 t2
+
+  (TyRecord _ ts1, TyRecord _ ts2) ->
+      -- records with the same field names, try unifying the field types
+      case mgus (M.elems ts1) (M.elems ts2) of
+        Right result -> Right result
+        Left msgs -> Left $ failMGUAdd msgs t1 t2
```

Comment:
I specifically didn't do that (here and below) in pursuit of clarity -- this way there's a single block for each case and each case begins with its condition.

Is writing ``` | otherwise``` on the following case with the same pattern an adequate compromise? That does not seem to generate noise from the compiler.

I've done this much in bf86a7f58c25f50c872b7d3455033fa784d7192e but I guess I'll adjust further if you insist.

Reply:
I'm really not a fan of the style in https://github.com/GaloisInc/saw-script/commit/bf86a7f58c25f50c872b7d3455033fa784d7192e at all. Writing `(TyRecord _ ts1, TyRecord _ ts2) | otherwise -> ...` (i.e., a case alternative with a single guard) like this is outright redundant, as the `otherwise` guard will always match.

My rationale for wanting a single case alternative with multiple guards is quite similar to your rationale, actually: I want there to be a single place in the code that you have to jump to when looking for how to handle `TyRecord`s. With the way it was written originally, there are two `(TyRecord ..., TyRecord ...)` case alternatives, which requires readers to carefully stare at them to tell to see if there are any subtle differences in how each alternative matches on its patterns. There aren't any such differences, so I usually favor combining the alternatives into one to reduce cognitive load.

I don't want to die on this hill, so I'd be fine with reverting back to how you had it originally if you feel strongly.

Reply:
I am used to reading cases this way because of years of OCaml (where you can only have one guard per case, plus there are other obscure ways the order of cases can become significant) so for me the critical part for legibility is being able to see the case patterns clearly.

But I don't want to die on this hill either, so I'll try it the other way...

(I don't see a way to do the one in inferStmts that doesn't end up looking like line noise, so I changed the guards there to an if/else)